wallet: Replace CDataStream& with CDataStream&& where appropriate

maflcko opened the pull request:
The keys and values are only to be used once because their memory is set
to zero. Make that explicit by moving the bytes into the lower level
methods.

sipa commented on src/wallet/bdb.h:
```diff
@@ -208,10 +208,10 @@ class BerkeleyBatch
-    bool WriteKey(CDataStream& key, CDataStream& value, bool overwrite=true);
-    bool EraseKey(CDataStream& key);
-    bool HasKey(CDataStream& key);
+    bool ReadKey(Span<char> key, CDataStream& value);
```
These Spans can be const char ones. That also has the advantage of supporting automatic conversion from temporaries.

MarcoFalke replied:
`SafeDbt` is used to assign `0` to the `char`s, which wouldn't work with `const char`

ryanofsky commented on src/wallet/bdb.h:
```diff
@@ -208,10 +208,10 @@ class BerkeleyBatch
-    bool WriteKey(CDataStream& key, CDataStream& value, bool overwrite=true);
-    bool EraseKey(CDataStream& key);
-    bool HasKey(CDataStream& key);
+    bool ReadKey(Span<char> key, CDataStream& value);
```
In commit "wallet: Replace CDataStream& with Span<char> where possible" (fafeffc6b23bea9addc091f69fe72440af185db8)

Might be worth mentioning in a comment here that memory_cleanse will be called on all these spans. I didn't realize this was happening, even after your previous comment in https://github.com/bitcoin/bitcoin/pull/19292#discussion_r442239967. You could imagine this leading to bugs if for example an EraseKey call was made after a ReadKey call with the same span, nothing would be erased because the key would be wiped.

MarcoFalke replied:
Does it even make sense to wipe the memory of bdb keys (not values)? I'd be surprised if the db-keys itself are encrypted, so I'd rather not add this comment.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/19320#discussion_r442490601

> Does it even make sense to wipe the memory of bdb keys (not values)? I'd be surprised if the db-keys itself are encrypted, so I'd rather not add this comment.

No I think it's bad behavior that could cause bugs like reading and writing and reading then erasing, or calling haskey before readkey all not to work. The point of a comment mentioning keys are wiped would be to document this surprising behavior and prevent bugs. But the behavior precedes this PR so definitely feel to ignore it

ryanofsky reviewed (approved):
Code review ACK fafeffc6b23bea9addc091f69fe72440af185db8

If anyone wants to do more cleanup like this I had a note to change the walletdb [`ReadKeyValue`](https://github.com/bitcoin/bitcoin/blob/c2bcb99c1d113ac090914e87b6f5676b5b55f82e/src/wallet/walletdb.cpp#L265) function to accept Span arguments as well. That'd require an additional change to support deserializing from Spans, such as tweaking [`VectorReader`](https://github.com/bitcoin/bitcoin/blob/c2bcb99c1d113ac090914e87b6f5676b5b55f82e/src/streams.h#L146) to accept a Span instead of a vector

achow101 commented:
ACK fafeffc6b23bea9addc091f69fe72440af185db8

Commit:
wallet: Remove confusing double return value ret+success

Also, remove redundant comments

Commit:
wallet: Replace CDataStream& with CDataStream&& where appropriate

The keys and values are only to be used once because their memory is set
to zero. Make that explicit by moving the bytes into the lower level
methods.

MarcoFalke commented:
> leading to bugs if for example an EraseKey call was made after a ReadKey call with the same span, nothing would be erased because the key would be wiped between calls.

Fixed by using the double ampersand trick

ryanofsky reviewed (approved):
Code review ACK fa8a341b88cabfd7f8d702db7cb9972b0804bf2a. Nice changes.

PR is rewritten since last review, so warning to others some previous discussion isn't relevant anymore (might have been a clearer to close this PR and open a new one). I still think it'd be good to replace all the cdatastreams with spans, but this would require externalizing memory_cleanse to be safe, so would a bigger change. More could be done here but this PR is a nice improvement.

sipa commented:
utACK fa8a341b88cabfd7f8d702db7cb9972b0804bf2a

jonatack reviewed (commented):
Nice refactoring.

post-merge ACK fa8a341b88cabfd7f8d702db7cb9972b0804bf2a
